### PR TITLE
Refactor cluster manager/nodes/shards

### PR DESCRIPTION
**Pull request by vangheem**
### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

File: nucliadb/nucliadb/common/maindb/utils.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    from nucliadb.common.maindb.local import LocalDriver
+
+    FILES = True
+except ImportError:  # pragma: no cover
+    FILES = False
+
+_DRIVER_UTIL_NAME = "driver"
+
+
+async def get_driver() -> Driver:
```
Comment (jotare) [inline nucliadb/nucliadb/common/maindb/utils.py:56]: (doubt) Why are we lazily creating maindb driver? Other utilities are created and set at startup time. Lazy creation adds dependency with settings.

**Review by jotare — COMMENTED** [review]
Great work! I like the simplification of abstractions and moving ORM to a common module.

I've made a couple of not super related things (some of them were already in the codebase), so feel free to ignore them.

File: nucliadb/nucliadb/common/cluster/abc.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+class AbstractIndexNode(metaclass=ABCMeta):
+    label: str = "index-node"
+
+    def __init__(
+        self,
+        *,
+        id: str,
+        address: str,
+        shard_count: int,
+        dummy: bool = False,
```
[the remark this answers is not in the document]
Comment (lferran) [reply nucliadb/nucliadb/common/cluster/abc.py:44]: agree with @jotare , but maybe better in a separate PR, as this one is already quite big?

File: nucliadb/nucliadb/common/maindb/utils.py (hunk 1 above)
Comment (vangheem) [reply nucliadb/nucliadb/common/maindb/utils.py:56]: I agree and think it's worth while to fix this as well.

File: nucliadb/nucliadb/common/cluster/index_node.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    _writer: Optional[NodeWriterStub] = None
+    _reader: Optional[NodeReaderStub] = None
+    _sidecar: Optional[NodeSidecarStub] = None
+
+    def _get_service_address(
+        self, port_map: dict[str, int], port: Optional[int]
+    ) -> str:
+        hostname = self.address.split(":")[0]
+        if port is None:
+            # For testing proposes we need to be able to have a writing port
```
Comment (lferran) [inline nucliadb/nucliadb/common/cluster/index_node.py:53]:
```suggestion
            # For testing purposes we need to be able to have a writing port
```

File: nucliadb/nucliadb/common/cluster/settings.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    chitchat_binding_host: str = "0.0.0.0"
+    chitchat_binding_port: int = 31337
+
+    logging_config: Optional[str] = None
+
+    node_writer_port: int = 10000
+    node_reader_port: int = 10001
+    node_sidecar_port: int = 10002
+
+    # Only for testing proposes
```
Comment (lferran) [inline nucliadb/nucliadb/common/cluster/settings.py:39]:
```suggestion
    # Only for testing purposes
```

File: nucliadb/setup.py [hunk 5]
```diff
@@ -79,7 +79,6 @@ def load_reqs(filename):
             "nucliadb-extract-openapi-reader = nucliadb.reader.openapi:command_extract_openapi",
             "nucliadb-extract-openapi-search = nucliadb.search.openapi:command_extract_openapi",
             "nucliadb-extract-openapi-writer = nucliadb.writer.openapi:command_extract_openapi",
-            "nucliadb-dataset-upload = nucliadb.train.upload:run",
```
Comment (lferran) [inline nucliadb/setup.py:82]: may this be in use by some other team/user?

**Review by lferran — APPROVED** [review]

**Pull request merged**